[Stretch] Support stretch and duration expressions for `Delay` instruction.

kevinhartman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Adds support for using classical expressions of type `Stretch` and `Duration` as the `duration` of a `Delay` instruction.

This is probably the final PR in the series 🎉. All the missing QPY stuff I'll be doing on the earlier PR branches, and will merge all of that plus fixes forward through to this.


### Details and comments
~Based on #13852. [More readable diff here](https://github.com/kevinhartman/qiskit/compare/circuit-stretch...kevinhartman:qiskit:stretchy-delay).~

#### To-do
- ~[ ] Release note.~ edit: I'd say this is already captured in previous release notes of this series
- [x] Support transpile end-to-end.
- [x] Testing.
- [x] QPY.

kevinhartman committed:
WIP

kevinhartman committed:
Add try_const to lift.

kevinhartman committed:
Try multiple singletons, new one for const.

kevinhartman committed:
Revert "Try multiple singletons, new one for const."

This reverts commit e2b32212533a52b30fecb700ed36a7f9e93a19a6.

kevinhartman committed:
Remove Bool singleton test.

kevinhartman committed:
Add const handling for stores, fix test bugs.

kevinhartman committed:
Fix formatting.

kevinhartman committed:
Remove Duration and Stretch for now.

kevinhartman committed:
Cleanup, fix const bug in index.

kevinhartman committed:
Fix ordering issue for types with differing const-ness.

Types that have some natural order no longer have an ordering
when one of them is strictly greater but has an incompatible
const-ness (i.e. when the greater type is const but the other
type is not).

kevinhartman committed:
Fix QPY serialization.

We need to reject types with const=True in QPY until it supports them.

For now, I've also made the Index and shift operator constructors
lift their RHS to the same const-ness as the target to make it
less likely that existing users of expr run into issues when
serializing to older QPY versions.

kevinhartman committed:
Make expr.Lift default to non-const.

This is probably a better default in general, since we
don't really have much use for const except for timing
stuff.

kevinhartman committed:
Revert to old test_expr_constructors.py.

kevinhartman committed:
Make binary_logical lift independent again.

Since we're going for using a Cast node when const-ness
differs, this will be fine.

kevinhartman committed:
Update tests, handle a few edge cases.

kevinhartman committed:
Fix docstring.

kevinhartman committed:
Remove now redundant arg from tests.

kevinhartman committed:
Add const testing for ordering.

kevinhartman committed:
Add const tests for shifts.

kevinhartman committed:
Add release note.

kevinhartman committed:
Add const store tests.

kevinhartman committed:
Add Float type to classical expressions.

kevinhartman committed:
Allow DANGEROUS conversion from Float to Bool.

I wasn't going to have this, but since we have DANGEROUS
Float => Int, and we have Int => Bool, I think this makes the
most sense.

kevinhartman committed:
Test Float ordering.

kevinhartman committed:
Improve error messages for using Float with logical operators.

kevinhartman committed:
Float tests for constructors.

kevinhartman committed:
Add release note.

kevinhartman committed:
Address lint, minor cleanup.

kevinhartman committed:
Add Duration and Stretch classical types.

A Stretch can always represent a Duration (it's just an expression
without any unresolved stretch variables, in this case), so we
allow implicit conversion from Duration => Stretch.

The reason for a separate Duration type is to support things like
Duration / Duration => Float. This is not valid for stretches in
OpenQASM (to my knowledge).

kevinhartman committed:
Add Duration type to qiskit.circuit.
Also adds support to expr.lift to create a value expression of
type types.Duration from an instance of qiskit.circuit.Duration.

kevinhartman committed:
Block Stretch from use in binary relations.

kevinhartman committed:
Add type ordering tests for Duration and Stretch.

Also improves testing for other types.

kevinhartman committed:
Test expr constructors for Duration and Stretch.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Implement operators +, -, *, /.

kevinhartman committed:
Add expression tests for arithmetic ops.

kevinhartman committed:
Implement stretch support for circuit.

kevinhartman committed:
Initial support for writing QASM3.

kevinhartman committed:
Support duration and stretch expressions in Delay.

kevinhartman committed:
Reject const vars in add_var and add_input.

Also removes the assumption that a const-type can never be an l-value
in favor of just restricting l-values with const types from being
added to circuits for now.

We will (in a separate PR) add support for adding stretch variables
to circuits, which are const. However, we may track those
differently, or at least not report them as variable when users
query the circuit for variables.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Implement QPY support for const-typed expressions.

kevinhartman committed:
Remove invalid test.

This one I'd added thinking I ought to block store from using
a const var target. But since I figured it's better to just
restrict adding vars to the circuit that are const (and leave
the decision of whether or not a const var can be an l-value
till later), this test no longer makes sense.

kevinhartman committed:
Update QPY version 14 desc.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Add serialization testing.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Merge branch 'const-expr' into expr-float

kevinhartman committed:
Test pre-v14 QPY rejects const-typed exprs.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Merge branch 'const-expr' into expr-float

kevinhartman committed:
QASM export for floats.

kevinhartman committed:
QPY support for floats.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Merge branch 'expr-float' into expr-timing

kevinhartman committed:
Settle on Duration circuit core type.

kevinhartman committed:
QPY serialization for durations and stretches.

kevinhartman committed:
Add QPY testing.

I can't really test Stretch yet since we can't add them to circuits
until a later PR.

kevinhartman committed:
QASM support for stretch and duration.

The best I can do to test these right now (before Delay
is made to accept timing expressions) is to use them in
comparison operations (will be in a follow up commit).

kevinhartman committed:
Fix lint.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Add arithmetic operators to QASM.

kevinhartman committed:
QPY testing for arithmetic operations.

kevinhartman committed:
QASM testing for arithmetic operations.

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Update tests for blocked const vars.

We decided to punt on the idea of assigning const-typed variables,
so we don't support declaring stretch expressions via add_var or
the 'declarations' argument to QuantumCircuit. We also don't
allow const 'inputs'.

kevinhartman committed:
Only test declare stretch QASM; fix lint.

kevinhartman committed:
Special case for stretch in add_uninitialized_var.

At the moment, we track stretches as variables, but these
will eventually make their way here during circuit copy
etc., so we need to support them even though they're
const.

kevinhartman committed:
Remove outdated docstring comment.

kevinhartman committed:
Remove outdated comment.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Support Delay duration expression in transpiler.

kevinhartman committed:
Improve docstring for Delay.

kevinhartman committed:
Error when stretch is used in a circuit during scheduling.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Merge branch 'const-expr' into expr-float

kevinhartman committed:
Merge branch 'expr-float' into expr-timing

kevinhartman committed:
Don't use match since we still support Python 3.9.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Block const stores.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Fix enum match.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Fix bad merge.

kevinhartman committed:
Add a few more serialization tests.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Add additional testing for Duration expression eval.

kevinhartman committed:
Fix missing round when converting expr to 'dt'.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Revert visitors.py.

kevinhartman committed:
Address review comments.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Improve type docs.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into const-expr

kevinhartman committed:
Revert QPY, since the old format can support constexprs.

By making const-ness a property of expressions, we don't need
any special serialization in QPY. That's because we assume that
all `Var` expressions are non-const, and all `Value` expressions
are const. And the const-ness of any expression is defined by
the const-ness of its operands, e.g. when QPY reconstructs a
binary operand, the constructed expression's `const` attribute
gets set to `True` if both of the operands are `const`, which
ultimately flows bottom-up from the `Var` and `Value` leaf nodes.

kevinhartman committed:
Move const-ness from Type to Expr.

kevinhartman committed:
Revert QPY testing, no longer needed.

kevinhartman committed:
Add explicit validation of const expr.

kevinhartman committed:
Revert stuff I didn't need to touch.

kevinhartman committed:
Update release note.

kevinhartman committed:
A few finishing touches.

kevinhartman committed:
Merge branch 'const-expr-expr' into expr-float-expr

kevinhartman committed:
Fix-up after merge.

kevinhartman committed:
Merge branch 'expr-float-expr' into expr-timing-expr

kevinhartman committed:
Fix-ups after merge.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Fix comment and release note.

kevinhartman committed:
Merge branch 'expr-float' into expr-timing

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Fixes after merge.

kevinhartman committed:
Fix test.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into expr-float

kevinhartman committed:
Merge branch 'expr-float' into expr-timing

kevinhartman committed:
Special-case Var const-ness for Stretch type.

This feels like a bit of a hack, but the idea is to override a
Var to report itself as a constant expression only in
the case that its type is Stretch. I would argue that it's not
quite as hack-y as it appears, since Stretch is probably the
only kind of constant we'll ever allow in Qiskit without an
in-line initializer. If ever we want to support declaring
other kinds of constants (e.g. Uint), we'll probably want to
introduce a `expr.Const` type whose constructor requires a
const initializer expression.

kevinhartman committed:
Address review comments.

kevinhartman committed:
Update release note.

kevinhartman committed:
Merge branch 'expr-float' into expr-timing

kevinhartman committed:
Update docs.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Add release notes and doc link.

kevinhartman committed:
Address review comments.

kevinhartman committed:
Merge branch 'expr-float' into expr-timing

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into expr-timing

kevinhartman committed:
Remove Stretch type.

kevinhartman committed:
Remove a few more mentions of the binned stretch type.

kevinhartman committed:
Add docstring for Duration.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Remove Stretch type stuff.

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
WIP

kevinhartman committed:
Track stretch variables throughout circuits.

kevinhartman committed:
Update QASM exporter.

kevinhartman committed:
Fix existing tests and found bugs.

kevinhartman committed:
Fix format.

kevinhartman committed:
Simplify structural eq visitor.

kevinhartman committed:
Add num_identifiers.

kevinhartman committed:
Support QPY.

kevinhartman committed:
Implement QASM visit for stretch expr.

kevinhartman committed:
Track stretches through DAGCircuit.

kevinhartman committed:
Add visit_stretch to classical resource map.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Some merge fixes.

kevinhartman committed:
Address review comments.

kevinhartman committed:
Remove unused import.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Represent stretch with StretchDeclaration in QASM AST.

Previously, we used StretchType within a ClassicalDeclaration,
but this wasn't the best fit.

kevinhartman committed:
Remove visitor short circuit.

kevinhartman committed:
Merge branch 'expr-timing' into math-expr

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Address review comments.

kevinhartman committed:
Address review comments.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into math-expr

kevinhartman committed:
Support division by Uint.

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Add release note.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Fix lint.

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Add test for QASM example.

kevinhartman committed:
Add missing DAG stretch plumbing.

kevinhartman committed:
Fix test.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Fix QPY serialization bug.

kevinhartman committed:
Fix lint.

kevinhartman committed:
Add qpy test.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Support remapping for stretch variables in compose.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into math-expr

kevinhartman committed:
Merge branch 'math-expr' into circuit-stretch

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Fix qpy compat test.

kevinhartman committed:
Add negative test for QPY stretch expr.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into circuit-stretch

kevinhartman committed:
Add more circuit testing.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Fix tests after merge.

kevinhartman committed:
Add control flow builder tests.

kevinhartman committed:
Add circuit equality testing for stretch.

kevinhartman committed:
Refer to 'stretches' instead of 'stretch variables'.

kevinhartman committed:
Remove | None for Stretch.name

jakelishman commented on test/python/compiler/test_transpiler.py:
```diff
@@ -1329,6 +1362,206 @@ def test_delay_converts_to_dt(self):
+
+    def test_rejects_mixed_units_delay_expr_without_target_dt(self):
+        """Test that a delay instruction with an expression duration is rejected
+        when the duration resolves to a negative number."""
```
Completely unimportant, but this docstring doesn't match.

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
+}
+
+
+class _EvalDurationImpl(expr.ExprVisitor[float]):
+    """Evaluates the expression to a single float result.
+
+    If ``dt`` is provided or all durations are already in ``dt``, the result is in ``dt``.
+    Otherwise, the result will be in seconds, and all durations MUST be in wall-time (SI).
+    """
+
+    __slots__ = ("dt", "has_dt", "has_si")
```
Absolutely applaud the effort here haha.  I'm not certain who's going to write an expression using only duration literals in a way that is valid through this pass via constant folding, but I love that it'll work if they do.

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
+    def __init__(self, dt=None):
+        self.dt = dt if dt is not None else 1
+        self.has_dt = False
+        self.has_si = False
```
Would leaving `self.dt = None` be unsafe here?  Feels like `1` is a technically valid value for `dt`.  Can't imagine that there are many QCs out there that have a cycle time of 1s, but still.

kevinhartman replied:
Yeah, this was sort of gross for me to have done lol.

Another "divide by 1" for less code trick, which I'll change to just be local rather than at the instance level.

60e6b5cc12db9a89599f3cd3e16dc3f51291a5fb

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
+                raise TranspilerError(
+                    "Fail to unify time units in delays. SI units "
+                    "and dt unit must not be mixed when dt is not supplied."
+                )
```
Technically I think the `self.dt == 1` check is unnecessary, since we both have SI and dt - we're going to be failing either way, right, even if dt isn't set?  (Not important to change.)

kevinhartman replied:
If we have both SI and cycles, we're fine since we can convert, unless we don't have a `self.dt`.

This might just be hard to read. I changed `self.dt` to be `None` in this case, so now this reads `if self.has_si and self.dt is None:`

jakelishman replied:
oh of course yeah - that's the whole purpose of the pass lol

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
+            from_unit = _DURATION_KIND_NAME.get(type(node.value))
+            return apply_prefix(node.value[0], from_unit) / self.dt
+        raise TranspilerError(f"invalid duration expression: {node}")
+
+    def visit_binary(self, node, /) -> float:
```
This makes me realise that we added the float type but didn't add unary negation as an op.  No need to do it in this PR, just noting that we should add it for Qiskit 2.1.

kevinhartman replied:
Ah that's a good point. Created issue #13966

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
+            self.has_si = True
+            if isinstance(node.value, Duration.s):
+                return node.value[0] / self.dt
+            from_unit = _DURATION_KIND_NAME.get(type(node.value))
```
I think we added `Duration.unit` that does this now?  (Not important to change, really.)

kevinhartman replied:
Very important, it's much cooler and makes that `unit` method feel even more useful 😉

60e6b5cc12db9a89599f3cd3e16dc3f51291a5fb

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
+        return node.operand.accept(self)
+
+    def visit_index(self, node, /) -> float:
+        return node.target.accept(self)
```
Probably we ought to leave this undefined and/or let it error out - it can't currently appear in valid expressions of type `Duration`, and this implementation wouldn't be valid (doesn't examine the index) for most hypothetical working implementations.

kevinhartman replied:
Done in 60e6b5cc12db9a89599f3cd3e16dc3f51291a5fb

kevinhartman committed:
Address review comments.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into circuit-stretch

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -71,52 +73,90 @@ def run(self, dag: DAGCircuit):
+                    duration = rounded_duration
+                else:
+                    has_si = True
+                if duration < 0:
+                    raise TranspilerError(
+                        f"Expression '{node.op.duration}' resolves to a negative duration!"
+                    )
```
I don't think we do this for legacy durations, even though they're representable.  Do we do the check somewhere else (and so should let that somewhere-else handle it), or is it a failure path we should handle for other things?

Fwiw, `TimeUnitConversion` _itself_ probably doesn't care if a duration is negative (the conversions are well defined anyway), but other places would.

kevinhartman replied:
This would normally blow up in the parameter validation for `Delay`. I can delete this check if we'd rather push down the error. I just figured it might be weird to get a `CircuitError` during transpilation, but I have no real reason to think that.

Edit: I think this probably _is_ `TimeUnitConversion`'s responsibility, since it'd be the thing creating the invalid `Delay` (we assign to `Delay.duration`, which end up calling parameter validation).

jakelishman replied:
Oh, if literals would blow up because `Delay` wouldn't accept it, it's fine.

kevinhartman replied:
Yup, yeah they would.

jakelishman commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -71,52 +73,90 @@ def run(self, dag: DAGCircuit):
+                        warnings.warn(
+                            f"Duration is rounded to {rounded_duration:d} [dt] from {duration:f}",
+                            UserWarning,
+                        )
```
Do you reckon it'd be more or less gross to call `duration_in_dt(duration, 1.0)` to get this effect?  I don't love my thing, but I also feel weird about duplicating the logic.

kevinhartman replied:
I actually thought to do this originally, but then I read the code inside that method wrong and thought this wouldn't work. But it does, so thanks! 😄

jakelishman reviewed (commented):
I think this is a complete review of the last PR.  I promised myself I wasn't going to comment significantly on the `TimeUnitConversion` as I think it's a bonus feature for 2.0, but then I went ahead and just did it anyway because I have no self control.

kevinhartman committed:
Update QPY desc.

kevinhartman committed:
Fix num_identifiers.

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into circuit-stretch

kevinhartman committed:
Fix merge for control flow tests.

kevinhartman committed:
Merge branch 'circuit-stretch' into stretchy-delay

kevinhartman committed:
Merge branch 'main' of github.com:Qiskit/qiskit into stretchy-delay

kevinhartman committed:
Address review comments.

kevinhartman committed:
Fix docstrings.

kevinhartman committed:
Use Union instead of | for QC delay.

Python 3.9 might not define the | operator for ParameterValueType since it's an alias.

jakelishman reviewed (dismissed):
Thanks for all this work - best as I could see here, this is ready to go!

mtreinish commented on qiskit/circuit/delay.py:
```diff
@@ -107,6 +132,12 @@ def validate_parameter(self, parameter):
+            if not parameter.const:
+                raise CircuitError("Duration expressions must be constant.")
+            return parameter
         elif isinstance(parameter, ParameterExpression):
```
It always surprises me that we allow this. Who is going to make their delay  `acos(A) / pi + e^(pi*B)`

mtreinish commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -4553,17 +4553,20 @@ def barrier(self, *qargs: QubitSpecifier, label=None) -> InstructionSet:
 
         Args:
-            duration (int or float or ParameterExpression): duration of the delay.
+            duration (Object):
```
FWIW you can just drop `(Object)` here it's all superseded by the type hinting in the docs.

mtreinish commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -71,52 +74,83 @@ def run(self, dag: DAGCircuit):
+                    raise TranspilerError(
+                        f"Expression '{node.op.duration}' resolves to a negative duration!"
+                    )
+                expression_durations[node._node_id] = duration
```
I feel like we had a potential gotcha around using node ids as indexes/keys like this that triggered a bug in the 1.2, 1.3 timeframe. It might have just been that they ids are reused after removals and aren't unique between DAGs. But nothing here would change the node index so we should be fine.

kevinhartman replied:
Yeah, I think this is actually the _only_ way to do this sort of thing, so I was very happy when I remembered that the node ID was accessible from Python. 

I know we had a "gotcha" of using the nodes directly in a map/set since they're ephemeral.

mtreinish replied:
Ah that was it, putting the node object itself in the set was what I was thinking of

mtreinish commented on qiskit/transpiler/passes/scheduling/time_unit_conversion.py:
```diff
@@ -71,52 +74,83 @@ def run(self, dag: DAGCircuit):
+                    # passes anyway, so we bail out. In theory, we _could_ still traverse
+                    # through the stretch expression and replace any Duration value nodes it may
+                    # contain with ones of the same units, but it'd be complex and probably unuseful.
+                    self.property_set["time_unit"] = "stretch"
```
Does anything use this property set value? If so how do they handle the string stretch here? I would expect our current scheduling passes to just skip or fail in the presence of stretches. 

Is this what raises the error when transpilation is called with a stretch duration?

kevinhartman replied:
Yeah, the scheduling passes look at this and raise if it's `"stretch"`. I think some of the other uses of `time_unit` in the property set have gone away since I started this work, but I think we still need something like this to let the scheduling passes know they shouldn't run, unless we want to examine all of the delays again.

mtreinish commented on test/python/compiler/test_transpiler.py:
```diff
@@ -1316,6 +1317,38 @@ def test_circuit_with_delay(self, optimization_level):
+        with self.assertWarnsRegex(
+            DeprecationWarning,
+            expected_regex="The `target` parameter should be used instead",
+        ):
```
This was removed in the most recent merge I think so you'll probably need to update again to account for the changes (and use a target).

mtreinish replied:
Oh yep, ci is failing now on the jobs that triggered after the PR merged: https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=65343&view=logs&j=8ec877a4-4914-5e33-8fb1-060b8cba69e2&t=414c1271-fa57-5a5a-9f2d-85b6cf71e8eb&l=18275

mtreinish reviewed (commented):
Overall this looks great, thanks for doing this! I left a few inline comments but none of them need fixing (most are just idle musings on my part) except for the tests which will cause a conflict because main has changed on you and you can't use the deprecated arguments to transpile you were using before.

kevinhartman committed:
Update tests for deprecation (sorry!)

kevinhartman committed:
Merge branch 'main'of github.com:Qiskit/qiskit into stretchy-delay

kevinhartman committed:
Trigger CI.
